source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,630 @@
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ ulIndex ];
+
+        if( pxContext->_internal.usMqttSubPacketID == usPacketId )
+        {
+            /* TODO: inspect packet to see is ack was successful*/
+            pxContext->_internal.usState = azureiothubTOPIC_SUBSCRIBE_STATE_SUBACK;
+            AZLogInfo( ( "Suback receive context found: 0x%08x", ulIndex ) );
+            break;
+        }
+    }
+
+    /* If reached the end of the list and haven't found a context, log none found */
+    if( ulIndex == azureiothubSUBSCRIBE_FEATURE_COUNT )
+    {
+        AZLogInfo( ( "No receive context found for incoming suback." ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Process incoming MQTT events.
+ *
+ * */
+static void prvEventCallback( AzureIoTMQTTHandle_t pxMQTTContext,
+                              AzureIoTMQTTPacketInfo_t * pxPacketInfo,
+                              AzureIoTMQTTDeserializedInfo_t * pxDeserializedInfo )
+{
+    /* First element in AzureIoTHubClientHandle */
+    AzureIoTHubClient_t * pxAzureIoTHubClient = ( AzureIoTHubClient_t * ) pxMQTTContext;
+
+    if( ( pxPacketInfo->ucType & 0xF0U ) == azureiotmqttPACKET_TYPE_PUBLISH )
+    {
+        prvMQTTProcessIncomingPublish( pxAzureIoTHubClient, pxDeserializedInfo->pxPublishInfo );
+    }
+    else if( ( pxPacketInfo->ucType & 0xF0U ) == azureiotmqttPACKET_TYPE_SUBACK )
+    {
+        prvMQTTProcessSuback( pxAzureIoTHubClient, pxPacketInfo, pxDeserializedInfo->usPacketIdentifier );
+    }
+    else
+    {
+        AZLogDebug( ( "AzureIoTHubClient received packet of type: 0x%08x", pxPacketInfo->ucType ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Check/Process messages for incoming Cloud to Device messages.
+ *
+ * */
+static uint32_t prvAzureIoTHubClientC2DProcess( AzureIoTHubClientReceiveContext_t * pxContext,
+                                                AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                void * pvPublishInfo )
+{
+    AzureIoTHubClientResult_t xResult;
+    AzureIoTHubClientCloudToDeviceMessageRequest_t xCloudToDeviceMessage = { 0 };
+    AzureIoTMQTTPublishInfo_t * xMQTTPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pvPublishInfo;
+    az_result xCoreResult;
+    az_iot_hub_client_c2d_request xOutEmbeddedRequest;
+    az_span xTopicSpan = az_span_create( ( uint8_t * ) xMQTTPublishInfo->pcTopicName, xMQTTPublishInfo->usTopicNameLength );
+
+    /* Failed means no topic match */
+    xCoreResult = az_iot_hub_client_c2d_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                              xTopicSpan, &xOutEmbeddedRequest );
+
+    if( az_result_failed( xCoreResult ) )
+    {
+        AZLogWarn( ( "Cloud to device topic parsing failed: 0x%08x", xCoreResult ) );
```

Comment:
I don't think this should be a warning - i don't think it should be logged at all, in fact.  This is going to be part of normal operation for say a deviceMethod invocation, and these warnings would cause a thoughtful developer to spend time trying to understand what was going wrong.

Reply:
I'd put in this block (and others) a brief note about how embedded API works because the topicNotMatch failure may not be obvious to someone debugging something that's not familiar with embedded layer.

Maybe something like in the comment right above embedded to expand a bit:
/* Failed means no topic match.  This is not a fatal error, but only means this message wasn't intended for C2D. */

Reply:
Yes in the full version there is a "process" function devoted to a specific subscription. So this embedded "parse" is supposed to only get the information out not necessarily check that it is for this feature. If this fails, it means somehow a topic got through which isn't for this subscibe.
https://github.com/Azure/azure-iot-middleware-freertos/blob/98b041e3cbf874a47e684451397f8603c8503a39/source/azure_iot_hub_client.c#L829

Will add the comment clarification